File: typescript/src/fetch.ts
/**
 * Drop-in fetch() wrapper that handles x402 payments automatically.
 *
 * Two ways to use it:
 *
 * 1. **Named wrapper** — inject into any SDK that accepts custom fetch:
 *    ```ts
 *    import { Wallet, createPayFetch } from "@pay-skill/sdk";
 *    const wallet = await Wallet.create();
 *    const payFetch = createPayFetch(wallet);
 *
 *    // Inject into OpenAI, Anthropic, Vercel AI SDK, etc.
 *    import OpenAI from "openai";
 *    const openai = new OpenAI({ fetch: payFetch });
 *    ```
 *
 * 2. **Global patch** — every fetch() in the process handles 402:
 *    ```ts
 *    import { Wallet, register } from "@pay-skill/sdk";
 *    const wallet = await Wallet.create();
 *    const unregister = register(wallet);
 *    ```
 *
 * @module
 */

import type { Wallet } from "./wallet.js";
import { PayBudgetExceededError } from "./errors.js";

// ── Types ─────────────────────────────────────────────────────────────

/** Budget and behavior options for createPayFetch. */
export interface PayFetchOptions {
  /**
   * Maximum dollars to pay for a single 402 settlement.
   * Requests exceeding this throw PayBudgetExceededError.
   * Default: no limit.
   */
  maxPerRequest?: number;

  /**
   * Maximum total dollars across all settlements in this wrapper's lifetime.
   * Default: no limit.
   */
  maxTotal?: number;

  /**
   * Called after each successful x402 payment.
   * Use for logging, observability, or UI updates.
   */
  onPayment?: (event: PaymentEvent) => void;
}

/** Metadata emitted after each successful x402 settlement. */
export interface PaymentEvent {
  /** The URL that required payment. */
  url: string;
  /** Dollar amount paid. */
  amount: number;
  /** How the payment was settled. */
  settlement: "direct" | "tab" | string;
}

// ── Implementation ────────────────────────────────────────────────────

/**
 * Create a fetch-compatible function that automatically settles
 * x402 (HTTP 402) responses using the provided wallet.
 *
 * The returned function has the exact same signature as `globalThis.fetch`.
 * Non-402 responses pass through untouched. When a 402 is received,
 * the wallet handles payment (tab or direct) and retries the request.
 *
 * @param wallet - A configured Wallet instance
 * @param options - Budget limits and callbacks
 * @returns A function with the same signature as fetch()
 *
 * @example Inject into OpenAI SDK
 * ```ts
 * import { Wallet, createPayFetch } from "@pay-skill/sdk";
 * import OpenAI from "openai";
 *
 * const wallet = await Wallet.create();
 * const payFetch = createPayFetch(wallet, { maxTotal: 10.00 });
 * const openai = new OpenAI({ fetch: payFetch });
 * ```
 *
 * @example Inject into Anthropic SDK
 * ```ts
 * import { Wallet, createPayFetch } from "@pay-skill/sdk";
 * import Anthropic from "@anthropic-ai/sdk";
 *
 * const wallet = await Wallet.create();
 * const client = new Anthropic({ fetch: createPayFetch(wallet) });
 * ```
 *
 * @example Standalone with budget controls
 * ```ts
 * const payFetch = createPayFetch(wallet, {
 *   maxPerRequest: 1.00,
 *   maxTotal: 50.00,
 *   onPayment: ({ url, amount, settlement }) => {
 *     console.log(`Paid $${amount} via ${settlement} for ${url}`);
 *   },
 * });
 *
 * const resp = await payFetch("https://api.example.com/data");
 * ```
 */
export function createPayFetch(
  wallet: Wallet,
  options: PayFetchOptions = {},
): typeof globalThis.fetch {
  const { maxPerRequest, maxTotal, onPayment } = options;
  let totalSpent = 0;

  const payFetch = async (
    input: string | URL | Request,
    init?: RequestInit,
  ): Promise<Response> => {
    // Make the initial request with the native fetch
    const resp = await fetch(input, init);
    if (resp.status !== 402) return resp;

    // Guard: don't retry if we already attached a payment (infinite loop)
    if (init?.headers) {
      const h = new Headers(init.headers);
      if (h.has("PAYMENT-SIGNATURE") || h.has("X-PAYMENT")) {
        return resp; // already tried paying, don't loop
      }
    }

    // Parse the 402 to check budget before committing
    const amountMicro = parse402Amount(resp);
    const amountDollars = amountMicro / 1_000_000;

    if (maxPerRequest !== undefined && amountDollars > maxPerRequest) {
      throw new PayBudgetExceededError(
        `Payment of $${amountDollars.toFixed(2)} exceeds per-request limit of $${maxPerRequest.toFixed(2)}`,
        totalSpent,
        amountDollars,
        "perRequest",
      );
    }

    if (maxTotal !== undefined && totalSpent + amountDollars > maxTotal) {
      throw new PayBudgetExceededError(
        `Payment of $${amountDollars.toFixed(2)} would exceed total budget of $${maxTotal.toFixed(2)} (spent: $${totalSpent.toFixed(2)})`,
        totalSpent,
        amountDollars,
        "total",
      );
    }

    // Normalize input to the args wallet.settle() needs
    const url =
      typeof input === "string"
        ? input
        : input instanceof URL
          ? input.toString()
          : input.url;

    // Serialize body to string if present (wallet.settle takes string)
    let bodyStr: string | undefined;
    if (init?.body != null) {
      bodyStr =
        typeof init.body === "string"
          ? init.body
          : init.body instanceof ArrayBuffer
            ? new TextDecoder().decode(init.body)
            : init.body instanceof Uint8Array
              ? new TextDecoder().decode(init.body)
              : String(init.body);
    }

    // Flatten headers to Record<string, string>
    const headers: Record<string, string> = {};
    if (init?.headers) {
      new Headers(init.headers).forEach((v, k) => {
        headers[k] = v;
      });
    }

    const result = await wallet.settle(resp, url, {
      method: init?.method,
      body: bodyStr,
      headers,
    });

    totalSpent += amountDollars;
    onPayment?.({
      url,
      amount: amountDollars,
      settlement: result.settlement,
    });

    return result.response;
  };

  return payFetch as typeof globalThis.fetch;
}

/**
 * Patch `globalThis.fetch` so every fetch() call in the process
 * automatically handles x402 payments.
 *
 * Returns an unregister function that restores the original fetch.
 *
 * @param wallet - A configured Wallet instance
 * @param options - Budget limits and callbacks
 * @returns A function that restores the original globalThis.fetch
 *
 * @example
 * ```ts
 * import { Wallet, register } from "@pay-skill/sdk";
 *
 * const wallet = await Wallet.create();
 * const unregister = register(wallet, { maxTotal: 50.00 });
 *
 * // Every fetch() call now handles 402 automatically
 * const resp = await fetch("https://api.example.com/data");
 *
 * // Restore original fetch when done
 * unregister();
 * ```
 */
export function register(
  wallet: Wallet,
  options?: PayFetchOptions,
): () => void {
  const original = globalThis.fetch;
  globalThis.fetch = createPayFetch(wallet, options);
  return () => {
    globalThis.fetch = original;
  };
}

// ── Internal ──────────────────────────────────────────────────────────

/**
 * Read the payment amount from a 402 response's headers without
 * consuming the response body (so wallet.settle can read it later).
 */
function parse402Amount(resp: Response): number {
  const header = resp.headers.get("payment-required");
  if (!header) return 0;
  try {
    const decoded = JSON.parse(atob(header)) as Record<string, unknown>;
    const accepts = decoded.accepts as
      | Array<Record<string, unknown>>
      | undefined;
    if (Array.isArray(accepts) && accepts.length > 0) {
      return Number(accepts[0].amount ?? 0);
    }
    return Number(decoded.amount ?? 0);
  } catch {
    return 0;
  }
}

File: python/src/payskill/fetch.py
"""Drop-in HTTP wrapper that handles x402 payments automatically.

Usage::

    from payskill import Wallet, create_pay_fetch

    wallet = Wallet()
    pay_fetch = create_pay_fetch(wallet, max_per_request=1.00, max_total=50.00)

    # Every call auto-pays 402 responses
    response = pay_fetch("https://api.example.com/data")

    # With httpx client (transport adapter)
    import httpx
    client = httpx.Client(transport=pay_fetch.transport())

Inject into any SDK that accepts custom HTTP clients::

    import anthropic
    client = anthropic.Anthropic(http_client=httpx.Client(transport=pay_fetch.transport()))
"""

from __future__ import annotations

import base64
import json
from collections.abc import Callable
from dataclasses import dataclass
from typing import TYPE_CHECKING, Any

import httpx

from payskill.errors import PayBudgetExceededError, PayNetworkError

# -- Types --------------------------------------------------------------------

if TYPE_CHECKING:
    from payskill.wallet import Wallet


@dataclass
class PayFetchOptions:
    """Budget and behavior options for create_pay_fetch."""

    max_per_request: float | None = None
    """Maximum dollars to pay for a single 402 settlement."""

    max_total: float | None = None
    """Maximum total dollars across all settlements in this wrapper's lifetime."""

    on_payment: Callable[[PaymentEvent], None] | None = None
    """Called after each successful x402 payment."""


@dataclass
class PaymentEvent:
    """Metadata emitted after each successful x402 settlement."""

    url: str
    """The URL that required payment."""

    amount: float
    """Dollar amount paid."""

    settlement: str
    """How the payment was settled ("direct" or "tab")."""


# -- Implementation -----------------------------------------------------------

_PAYMENT_HEADERS = frozenset({"payment-signature", "x-payment"})


class PayFetch:
    """Callable HTTP wrapper that auto-settles x402 responses.

    Call directly like a function, or use ``.transport()`` to get an
    httpx transport adapter for injection into httpx.Client.
    """

    def __init__(self, wallet: Wallet, options: PayFetchOptions | None = None) -> None:
        self._wallet = wallet
        opts = options or PayFetchOptions()
        self._max_per_request = opts.max_per_request
        self._max_total = opts.max_total
        self._on_payment = opts.on_payment
        self._total_spent: float = 0

    @property
    def total_spent(self) -> float:
        """Total dollars spent through this wrapper."""
        return self._total_spent

    def __call__(
        self,
        url: str,
        *,
        method: str = "GET",
        body: Any = None,
        headers: dict[str, str] | None = None,
    ) -> httpx.Response:
        """Make an HTTP request, auto-paying any 402 response.

        Args:
            url: The URL to request.
            method: HTTP method (default GET, POST if body provided).
            body: Request body (dict or string). Dicts are JSON-serialized.
            headers: Additional request headers.

        Returns:
            The final httpx.Response (after payment if 402 was encountered).

        Raises:
            PayBudgetExceededError: If payment would exceed budget limits.
            PayNetworkError: If the request fails due to network issues.
        """
        if body is not None and method == "GET":
            req_method = "POST"
        else:
            req_method = method
        req_headers = dict(headers) if headers else {}
        body_str = json.dumps(body) if isinstance(body, dict) else body

        try:
            resp = self._wallet._client.request(
                req_method,
                url,
                content=body_str.encode() if body_str else None,
                headers=req_headers,
            )
        except httpx.HTTPError as e:
            raise PayNetworkError(str(e)) from e

        if resp.status_code != 402:
            return resp

        # Don't retry if we already attached a payment header (prevent loops)
        lower_headers = {k.lower() for k in req_headers}
        if lower_headers & _PAYMENT_HEADERS:
            return resp

        # Parse amount from 402 to check budget before paying
        amount_micro = _parse_402_amount(resp)
        amount_dollars = amount_micro / 1_000_000

        if self._max_per_request is not None and amount_dollars > self._max_per_request:
            raise PayBudgetExceededError(
                f"Payment of ${amount_dollars:.2f} exceeds per-request limit "
                f"of ${self._max_per_request:.2f}",
                spent=self._total_spent,
                requested=amount_dollars,
                limit_type="per_request",
            )

        if self._max_total is not None and self._total_spent + amount_dollars > self._max_total:
            raise PayBudgetExceededError(
                f"Payment of ${amount_dollars:.2f} would exceed total budget "
                f"of ${self._max_total:.2f} (spent: ${self._total_spent:.2f})",
                spent=self._total_spent,
                requested=amount_dollars,
                limit_type="total",
            )

        # Settle payment and retry
        reqs = self._wallet._parse_402(resp)
        if reqs["settlement"] == "tab":
            result = self._wallet._settle_via_tab(url, req_method, body_str, req_headers, reqs)
        else:
            result = self._wallet._settle_via_direct(url, req_method, body_str, req_headers, reqs)

        self._total_spent += amount_dollars
        if self._on_payment is not None:
            self._on_payment(
                PaymentEvent(
                    url=url,
                    amount=amount_dollars,
                    settlement=reqs["settlement"],
                )
            )

        return result

    def transport(self) -> PayFetchTransport:
        """Create an httpx transport adapter for injection into httpx.Client.

        Usage::

            import httpx
            client = httpx.Client(transport=pay_fetch.transport())
            resp = client.get("https://api.example.com/data")
        """
        return PayFetchTransport(self)


class PayFetchTransport(httpx.BaseTransport):
    """httpx transport adapter that routes requests through PayFetch."""

    def __init__(self, pay_fetch: PayFetch) -> None:
        self._pay_fetch = pay_fetch

    def handle_request(self, request: httpx.Request) -> httpx.Response:
        headers = dict(request.headers)
        body_bytes = request.read()
        body_str = body_bytes.decode() if body_bytes else None

        return self._pay_fetch(
            str(request.url),
            method=request.method,
            body=body_str,
            headers=headers,
        )


# -- Public factory -----------------------------------------------------------


def create_pay_fetch(
    wallet: Wallet,
    *,
    max_per_request: float | None = None,
    max_total: float | None = None,
    on_payment: Callable[[PaymentEvent], None] | None = None,
) -> PayFetch:
    """Create an HTTP callable that automatically settles x402 responses.

    The returned object is callable with the same interface as a simple
    HTTP function. It also exposes ``.transport()`` for httpx integration.

    Args:
        wallet: A configured Wallet instance.
        max_per_request: Maximum dollars for a single payment.
        max_total: Maximum total dollars across all payments.
        on_payment: Callback after each successful settlement.

    Returns:
        A PayFetch instance (callable).

    Example::

        from payskill import Wallet, create_pay_fetch

        wallet = Wallet()
        pay = create_pay_fetch(wallet, max_per_request=1.00, max_total=50.00)
        resp = pay("https://api.example.com/data")

    Example with httpx.Client::

        import httpx
        from payskill import Wallet, create_pay_fetch

        wallet = Wallet()
        pay = create_pay_fetch(wallet)
        client = httpx.Client(transport=pay.transport())
        resp = client.get("https://api.example.com/data")
    """
    return PayFetch(
        wallet,
        PayFetchOptions(
            max_per_request=max_per_request,
            max_total=max_total,
            on_payment=on_payment,
        ),
    )


# -- Internal -----------------------------------------------------------------


def _parse_402_amount(resp: httpx.Response) -> int:
    """Extract payment amount from a 402 response without consuming the body."""
    header = resp.headers.get("payment-required")
    if not header:
        return 0
    try:
        decoded = json.loads(base64.b64decode(header))
        accepts = decoded.get("accepts")
        if isinstance(accepts, list) and len(accepts) > 0:
            return int(accepts[0].get("amount", 0))
        return int(decoded.get("amount", 0))
    except Exception:  # noqa: S110
        return 0
